tensorflow_io/avro/20190307-avro-dataset.md

```diff
@@ -0,0 +1,299 @@
+- Full support for N-dimensional dense tensors and N-dimensional sparse tensors
+- A mapping of Avro primitive types to TensorFlow primitive types (see Table 1)
+- For Avro&#39;s null type the default shall be assumed if provided; otherwise fail
+- An API/syntax that follows the one in Spark/Avro tooling that allows to resolve
+  - Elements in an array
+  - Elements in a map
+  - Nested records
+  - Namespaces
+  - Branches in unions -- will resolve them but will use strict typing which will render them useless whenever an attribute has multiple types or whenever a null is encountered. Note, TensorFlow uses strict typing for it&#39;s tensors.
+- An API for filtering in arrays
+- The mapping of Avro enumerations to TensorFlow strings carrying the symbol name
+- The mapping of Avro fixed types to TensorFlow strings
+- The same operating systems as supported by TensorFlow -- Note, that this only accounts for the development of the avro dataset and not it&#39;s dependencies
+- Python 2.x and 3.x following TensorFlow&#39;s requirements
+- Thread safety; Note that TensorFlow&#39;s  dataset API may use multiple threads
+
+| **Avro primitive type** | **TensorFlow primitive type** |
+| --- | --- |
+| null: no value | If provided default is assumed otherwise fail |
+| boolean: a binary value | tf.bool |
+| int: 32-bit signed integer | tf.int32 |
+| long: 64-bit signed integer | tf.in64 |
+| float: single precision 32-bit IEEE floating point number | tf.float32 |
+| double: double precision 64-bit IEEE floating point number | tf.float64 |
+| fixed: a fixed length byte sequence | tf.string |
+| bytes: a sequence of 8-bit unsigned bytes | tf.string |
+| string: unicode character sequence | tf.string |
+| enum: enumeration | tf.string |
+
+**Table 1** lists the mapping of Avro primitive types to TensorFlow primitive types.
+
+The Avro dataset shall NOT support
+
+- --Type promotion or type coercion when mapping from Avro types to TensorFlow types, instead we use strict typing here
+- --Recursive type definitions in avro
+- --The compile and distribution of TensorFlow external tools that are necessary; e.g. libavro.so. It assumed that the user will compile and distribute this library separately.
+
+## Design Details
+
+The design shall leverage existing software as follows
+
+- For the API leverage the tf.dataset natively and corresponding python classes
+- Leverage the TensorFlow filesystem natively. Note, that currently meta information for file systems are mostly provided through environment variables (e.g. HDFS). For more information follow the special interest group (SIG) tensorflow.io and these links: [Hadoop](https://www.tensorflow.org/deploy/hadoop) and [gfile](https://www.tensorflow.org/api_docs/python/tf/gfile) .
+- Leverage Avro&#39;s c API. Note, the c++ API for Avro depends on boost which the TensorFlow author&#39;s do not want to depend on nor maintain
```

Comment:
In tensorflow-io we include the boost already (for parquet and arrow). We could do link against boost or c. Either way is fine, no hard requirement I think.

Reply:
Let's leave it as c for now. For the avro reader I've encapsulated that in it's own class and it can easily be changed in the future should the need arise.  For the avro value I did not see a need for encapsulation for now.

Reply:
Since we include boost already, I would prefer reusing it since C++ API might be easier to maintain but open to other approach as well if it's easier to implement.

Reply:
@terrytangyuan Can you share the commit(s)/PR which includes the build of boost here?  Thanks

Reply:
@fraudies The boost library is included in:
https://github.com/tensorflow/io/blob/8c34c19d45598f6e881cc0bdc8820b831400405e/WORKSPACE#L105-L115